### DISCUSSION TITLE
How to bundle / serve multiple entrypoints not-relative to top-level project via command-line?

### DISCUSSION THREAD

**Discussion by nt3rp** [parcel-bundler/parcel#6277] [Q&A] [answered]
I have a project that resembles this:

<details>
  <summary>Code</summary>

```
+ project
    + src
        + graphics
            - graphic1.html
            - graphic1.js
            - graphic2.html
            - graphic2.js
         + dashboard
            - dashboard1.html
            - dashboard1.js
            - dashboard2.html
            - dashboard2.js
```
</details>

What I want is output like this:

<details>
  <summary>Code</summary>

```
+ project
    + graphics
        - graphic1.html
        - graphic1.<hash>.js
        - graphic2.html
        - graphic2.<hash>.js
     + dashboard
        - dashboard1.html
        - dashboard1.<hash>.js
        - dashboard2.html
        - dashboard2.<hash>.js
    + src
        + graphics
        + dashboard
```
</details>

I can get partway there with `--dist-dir` and glob...

```sh
npx parcel serve './src/**/*.html' --dist-dir=. --public-url=.
```

... but then if I look at the file paths in the generated HTML, the script tags are all relative to the project directory. For example:

```html
<!-- want this to be just `alert.<hash>.js` -->
<script src="graphics/graphic1.<hash>.js" />
```

Through experimentation, I've discovered that this will _build_ the files in the right place with the right path:

```
npx parcel build './src/graphics/*.html' --dist-dir=graphics --public-url=.
```

...But then I need to run that for every folder in my app, and serving the files behaves strangely.

**Any suggestions?** _Any_ help would be appreciated 🙇 

I've been grepping some of the source code and trying to wrap my head around the different concepts (i.e. packaging, bundling, namers, etc.) and I haven't had a lot of luck. I also read about the [`source` field](https://v2.parceljs.org/configuration/package-json/#source) but it [isn't able to accept blobs](https://github.com/parcel-bundler/parcel/issues/3821)? 😵

I also found [this similar question for a solution in Parcel v1](https://github.com/parcel-bundler/parcel/issues/1771) but it is quite old and the author didn't find a solution.

**Comment by nt3rp** [comment]
After some experimenting, I managed to [adapt someone else's partial solution](https://github.com/parcel-bundler/parcel/issues/1771) using the Parcel API.

Including the solution here in case others are trying to make something like this work (in my case, for NodeCG):

  ```nodejs
  // Based on this previous solution:
  // https://github.com/parcel-bundler/parcel/issues/1771
  const appRoot = require("app-root-path");
  const Parcel = require("@parcel/core").default;
  const parcelConfig = appRoot.resolve(".parcelConfig");
  const mode = process.argv[2];

  const runParcel = async ({ source, distDir, target = "browser" }) => {
    const bundler = new Parcel({
      entries: appRoot.resolve(source),
      defaultConfig: parcelConfig,
      defaultTargetOptions: {
        distDir,
        publicUrl: "./",
      },
    });

    switch (mode) {
      case "build":
        await bundler.run();
        break;
      case "watch":
        // TODO: Output changes in watch mode
        await bundler.watch();
      break;
      default:
        console.log("Invalid mode");
    }
  };

  const entries = [
    {
      source: "src/graphics/*.html",
      distDir: "graphics",
    },
    {
      source: "src/dashboard/*.html",
      distDir: "dashboard",
    },
  ];

  for (const entry of entries) {
    runParcel(entry);
  }
  ```

Would still love to know if there's a better / easier way to do this with just the CLI, since it's much more full-featured and provides more output (my current build script provides no output at the moment).

**Reply by DeMoorJasper** [reply]
Fairly certain you want to use targets https://v2.parceljs.org/getting-started/configuration/#targets

**Reply by nt3rp** [reply]
I _think_ that's the right approach as well, but it doesn't seem to work with wildcards. For example, if I add the following to my `package.json`:

```json
  "graphics": "graphics/*.html",
  "targets": {
    "graphics": {
      "source": "src/graphics/*.html",
      "distDir": "graphics",
      "publicUrl": "."
    }
  },
```

...And then run `npx parcel build .`, I get this error:

```
🚨 Build failed.
unknown: src/graphics/*.html in package.json#targets["graphics"].source does not exist
```

It looks like I can use targets if I write them like this:

```json
"graphics": "graphics/alert.html",
  "targets": {
    "graphics": {
      "source": "src/graphics/alert.html",
      "distDir": "graphics",
      "publicUrl": "."
    }
  },
```

But then I have to write a blob like that for every HTML file. I was hoping there was a way to do this without all the boilerplate especially since `npx parcel build './src/**/*.html' --dist-dir=. --public-url=.` is so close to doing just that.

**Reply by DeMoorJasper** [reply]
Yes globs aren't supported this is partially because npm doesn't support this.

For this you would use `parcel serve './src/**/*.html' --dist-dir=. --public-url=.` as you mentioned and probably use a custom namer? We still need to make the namer more configurable at some point (without having to write any code)

**Reply by nt3rp** [reply]
Ok, I'll take a look into that. Thanks for your help! 😀

**Reply by nt3rp** [reply]
I'm not having any luck with the naming route, but I think I have some more information. While digging through the source code I came across `entryRoot`. From what I can tell, it isn't ever set anywhere, and it cannot be set from anywhere (it can only be set as `InitialParcelOptions`, which doesn't include it, or as `PluginOptions`, though I can't find an example of where Plugins are configured with options).

Further, the docs have this [note on module resolution](https://v2.parceljs.org/features/module-resolution/): 

> **root directory**: the directory of the entrypoint specified to Parcel, or the shared root (nearest common parent directory) when multiple entrypoints are specified.

It looks like [Parcel always defaults to the root directory](https://github.com/parcel-bundler/parcel/blob/ba57b6501632703d68923b214b0c0f6d47a34b1f/packages/core/core/src/resolveOptions.js) as a result.

Going to keep investigating but any other tips are welcome.

**Reply by nt3rp** [reply]
Nevermind, `entryRoot` is definitely not what I'm looking for either. If I somehow set `entryRoot` correctly, then the bundle is no longer put in folders but instead is put in the root directory :(

**Comment by nt3rp** [comment, marked as answer]
I ended up using [`concurrently`](https://github.com/kimmobrunfeldt/concurrently) to do this.

In `package.json`:

```json
{
  "scripts": {
    "watch:graphics": "npx parcel watch 'src/graphics/*.html' --dist-dir='graphics' --public-url='.'",
    "build:graphics": "npx parcel build 'src/graphics/*.html' --dist-dir='graphics' --public-url='.'",
    "watch:dashboard": "npx parcel watch 'src/dashboard/*.html' --dist-dir='dashboard' --public-url='.'",
    "build:dashboard": "npx parcel build 'src/dashboard/*.html' --dist-dir='dashboard' --public-url='.'"
  },
}
```

Then I can run the following to have everything generated in the right directory:

```shell
npx concurrently 'npm:watch*'
```
